Added PDLL to mpl_plugins (rewrite pattern for various dialects)

IanWood1 opened the pull request:
Added pdll linalg conversion code (compiles to pdl) that performs match and rewrite. Also modified the CMakeLists.txt to add test cases for linalg.


Todo
- [x] Add pdll support for the tosa/torch test cases
- [x] Remove build of pdll file from mmake, have ctest/filecheck build the file

IanWood1 committed:
squash all commits to remove unnamed contributor

IanWood1 commented on samples/custom_dispatch/cpu/mlp_plugin/mlp_linalg_pdll.mlir:
```diff
@@ -0,0 +1,116 @@
+// RUN: mlir-pdll %p/mlp_all_specs.pdll -x=mlir -I%S/../../../..//third_party/torch-mlir/include -I%S/../../../..//samples/custom_dispatch/cpu/mlp_plugin -I%S/../../../..//third_party/llvm-project/mlir/include -I%S/../../../../-build/llvm-project/tools/mlir/include > %t && \ 
+// RUN: iree-compile --iree-preprocessing-pdl-spec-filename=%t %s --compile-to=preprocessing | FileCheck %s
```
Is there a better way to do this? Not a fan of the relative paths

MaheshRavishankar reviewed (changes requested):
High level comments
1) We probably want to keep the pdll scripts separate... just like the pdl scripts are separate.
2) We dont want to use `mlir-pdll` this way... For one you can use `IREE_BINARY_DIR` to get the paths from there. Let me look around a bit more here.

ScottTodd commented on samples/custom_dispatch/cpu/mlp_plugin/CMakeLists.txt:
```diff
@@ -108,4 +113,65 @@ iree_lit_test_suite(
+  NAME
+    mlp_pdll_torch_example
+  SRCS
+    "mlp_torch_pdll.mlir"
```
Tests using input dialects (torch, tosa, stablehlo) should be only defined if their matching CMake option is set. Otherwise, developers who disable an input dialect will get unactionable test failures.

Here are the available options: https://github.com/openxla/iree/blob/a2476ceef53f6d4ec0f7a9cf36892647626d324b/CMakeLists.txt#L418-L420

Here is an example of how to use one:
https://github.com/openxla/iree/blob/a2476ceef53f6d4ec0f7a9cf36892647626d324b/compiler/bindings/python/test/tools/CMakeLists.txt#L18-L25

ScottTodd commented on samples/custom_dispatch/cpu/mlp_plugin/CMakeLists.txt:
```diff
@@ -108,4 +113,65 @@ iree_lit_test_suite(
+file(APPEND ${CMAKE_BINARY_DIR}/pdll_compile_commands.yml
+  "--- !FileInfo:\n"
+  "  filepath: \"${CMAKE_CURRENT_SOURCE_DIR}/mlp_all_specs.pdll\"\n"
+  "  includes: \"${TORCH_MLIR_ROOT_DIR}/include;${CMAKE_CURRENT_SOURCE_DIR};${MLIR_INCLUDE_DIRS}\"\n"
```
I'm quite skeptical of the references to `TORCH_MLIR_ROOT_DIR` in this PR. What is that used for? We should be defining a stable interface on files distributed with releases, and I can't easily see what this PDLL code actually depends on.

IanWood1 replied:
removed

ScottTodd commented on samples/custom_dispatch/cpu/mlp_plugin/CMakeLists.txt:
```diff
@@ -108,4 +113,65 @@ iree_lit_test_suite(
+    FileCheck
+    iree-compile
+    iree-run-module
+    mlir-pdll
```
We don't distribute mlir-pdll with IREE. Can this use something we already distribute instead? I see a `--iree-preprocessing-pdl-spec-filename=` flag used in some other tests.

We could look into distributing mlir-pdll or some iree-pdll, but I'd need to see more information about the architecture and data flows involved here. The tablegen deps in `samples/custom_dispatch/cpu/mlp_plugin/mlp_all_specs.pdll` scare me... anything user facing shouldn't be touching compiler source files. Is this a compiler developer feature and not a user feature? Does the https://github.com/openxla/iree/tree/main/samples/custom_dispatch README cover this usage already? Could also document in the PR description.

IanWood1 replied:
Yes, I agree. I talked with Mahesh and we decided on adding `--iree-preprocessing-pdll-spec-filename=`, which mirrors the flags for pdl and transform dialect. Is this in line with what you are thinking?

From my understanding, this is a compiler developer feature but @MaheshRavishankar might have a better answer. Also, the readme does not contain documentation which I can definitely add (I did add a readme to the `mlp_plugin` folder).

I have removed all the tablegen  include files since they aren't required when going from PDLL to PDL. However, the underlying LLVM PDLL utilities use the includes to preform error checking. So, if this is something people are going to spend time using It is probably worth having some way to add includes and tell iree where to find them.

ScottTodd commented on samples/custom_dispatch/cpu/mlp_plugin/mlp_all_specs.pdll:
```diff
@@ -0,0 +1,119 @@
+#include "mlir/Dialect/Tensor/IR/TensorOps.td"
+#include "mlir/Dialect/Tosa/IR/TosaOps.td"
+#include "torch-mlir/Dialect/Torch/IR/TorchOps.td"
+#include "torch-mlir/Dialect/TorchConversion/IR/TorchConversionOps.td"
```
Can we avoid mixing TOSA and Torch in a single file?

IanWood1 replied:
Done, includes have been removed and each rewrite is split into its individual file based on dialect

IanWood1 committed:
start on README

IanWood1 committed:
save

IanWood1 committed:
Merge branch 'main' into pdll_branch_2

bjacob committed:
[CPU][NFC] CPU/KernelDispatch cleanups (#17124)

A few things bothered me while working on #17115, splitting these fixes
into their own PR:
1. `splitParallelAndReductionTiles` takes an input-output parameter
`parallelTileSizes` that is a misnomer for its input role, where it
really means ALL tile sizes. Its name is only appropriate for its output
role. This naming problem was leaking to its call sites. Introducing a
separate local variable at call site to make this issue better
contained.
2. `setX86VectorTileSizes` was not x86-specific (anymore). It was also
taking `numLoops` redundantly with `genericOp`, and was unnecessarily
restricted to `linalg::GenericOp` when all it needed was a `LinalgOp`.
3. In `setMatmulRootConfig`, some in-place mutation of `vecTileSizes`
made the logic hard to follow. Introducing a separate
`inputVecTileSizes` to make this a bit clearer.

banach-space committed:
[CodeGen] Add a basic unit test for the TilingConfig class (#17072)

This is just the initial set-up - I will be adding more tests in
follow-up PRs.

Commit:
[runtime][cts] Add test to wait multiple times for the same semaphore value (#17125)

IanWood1 committed:
add iree command line support for pdll

ScottTodd committed:
Add Colab notebook showing Hugging Face import via Turbine. (#17093)

This shows a minimal import from Hugging Face using the style of code
from
https://github.com/nod-ai/SHARK-TestSuite/tree/main/e2eshark/pytorch/models.
I'd like to expand on this with other advanced export APIs (e.g. using
parameter files like Safetensors/GGUF/IRPA, Turbine-LLM, and custom
kernels), but figured I'd start with something fairly well tested and
simple.

The test case I forked from also used a text prompt for an audio -> text
model, so this just shows some numbers as output... I'd like to show
text or images as output, though that needs some interfacing
(WhisperProcessor wraps WhisperFeatureExtractor and WhisperTokenizer,
for example).

IanWood1 committed:
moved pdll tests/files to subfolder and fixed cmake issues

ScottTodd committed:
Harden how ConstEval uses llvm-cpu and the runtime libraries. (#17075)

* Fixed https://github.com/openxla/iree/issues/17070 by updating the
CMake options needed for ConstEval
* Replaced `IREE_CHECK_OK` usage with error handling
* Refactored `test/jit_globals.mlir`, adding coverage for llvm-cpu
(since that is actually running by default)
* I tried to keep all test cases in one file, but `--verify-diagnostics`
isn't compatible with that style of lit testing AFAICT
* This uncovered some bugs in https://github.com/openxla/iree/pull/16321
/ missing support for i4 types

IanWood1 committed:
clean up pdll code

IanWood1 committed:
small changes to readme

IanWood1 committed:
small changes to readme

IanWood1 committed:
readme change [skip ci]

IanWood1 committed:
add corresponding tosa tests + add comments to pdll files to match pdl

IanWood1 committed:
bazel format

IanWood1 committed:
lint script

MaheshRavishankar commented on 49:
```diff
@@ -0,0 +1 @@
+
```
Errant file?

MaheshRavishankar commented on compiler/src/iree/compiler/Preprocessing/Common/ApplyPDLPatterns.cpp:
```diff
@@ -21,9 +21,16 @@
 #include "mlir/Dialect/Tensor/IR/Tensor.h"
 #include "mlir/Dialect/Utils/StaticValueUtils.h"
 #include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/PDLPatternMatch.h.inc"
```
Do we need this?

MaheshRavishankar commented on compiler/src/iree/compiler/Preprocessing/Common/ApplyPDLPatterns.cpp:
```diff
@@ -566,6 +574,84 @@ class ApplyPDLPatternsPass
+
+namespace {
+
+class ApplyPDLLPatternsPass
```
We can move this pass to a separate file?

nirvedhmeshram replied:
In this case, we will need to move all the common functions to a utility header.

IanWood1 replied:
Thank you nirvedh, I wasn't sure what was preferred (one file vs seperate util, pdl, and pdll) since PDLL required functions private to the translation unit. Should be a quick change!

MaheshRavishankar commented on samples/custom_dispatch/cpu/mlp_plugin/CMakeLists.txt:
```diff
@@ -56,39 +56,43 @@ iree_lit_test_suite(
-    "driver=local-sync"
-    "hostonly"
-)
+if(IREE_INPUT_TOSA)
```
Oh, I didnt know we needed to do this. Good catch!

MaheshRavishankar reviewed (changes requested):
Fantastic!! This is exactly what I was hoping for!

 I just have a few minor comments.

Groverkss committed:
[VectorDistribution] Add distribution pattern for vector::MultiDimReductionOp (#17076)

This patch adds support for distribution vector::MultiDimReductionOp.
The distribution works in two steps:

- Local reduction: add a vector.multi_reduction to reduce across
batches, outers, element
- Subgroup reduction: Use gpu.shuffle to reduce among threads

nirvedhmeshram commented on compiler/plugins/input/TOSA/InputConversion/test/BUILD.bazel:
```diff
@@ -17,6 +17,7 @@ iree_lit_test_suite(
     srcs = enforce_glob(
         [
             "apply_pdl_patterns_tosa.mlir",
+            "apply_pdll_patterns_tosa.mlir",
```
I think we are missing the torch counterpart of this test?

nirvedhmeshram reviewed (commented):
LGTM overall, thanks.

bjacob committed:
Executable library call hooks system, and a sample Linux/CPU event implementation (#15803)

There are two parts to this PR:
* The `runtime/` changes bring a minimal weak-hook extension point
around executable library calls. It's behind a C-internal flag: to
enable it, configure with `cmake
-DCMAKE_C_FLAGS=-DIREE_HAL_EXECUTABLE_LIBRARY_CALL_HOOK .` . When
enabled, it links to weak symbols and calls them when they are defined.
The symbols are `iree_hal_executable_library_call_hook_{begin,end}`.
* The `experimental/` changes bring one such hook implementation,
building a shared library defining such
`iree_hal_executable_library_call_hook_{begin,end}` symbols, doing Linux
event counter queries, with a number of events defined, imported from
JSON tables in the Linux source tree. At the moment it only defines
AMD-specific events and generic events, but more could be added. See the
README.md file there for how it's used.

Context:
This is a clean-up of the tools we had developed last fall to study
Llama2 on Zen4 CPUs. It's the most meaningful use of event counters that
I've been able to make yet. The difficulty with these is always to find
a way to make an apples-to-apples comparison. What we found in that
Llama2 work last fall was that some critical dispatch function had a
fluctuating latency, as shown graphically in Tracy histograms, which
raised the question: if this dispatch function has sometimes higher and
sometimes lower latency, can we understand that so that we could perhaps
more often fall in the lower-latency case? The next step was to
correlate these fluctuating latencies with some specific event counts.
This is what this tool is about.

Commit:
[runtime][cts] Add test waiting on a semaphore for finite time and fix Vulkan driver (#17126)

Fix Vulkan semaphore wait timeout if the timeout has already
passed. Don't check for a timeout on the host side.
Let Vulkan handle the timeout.
If the timeout has passed and the semaphore value is big enough,
we want an OK status to be returned.

Disable semaphore_test for Metal.

IanWood1 committed:
fix markdownlint errors (squashed)

IanWood1 committed:
remove errant file

IanWood1 committed:
seperate pdl and pdll and move fns to utility file

IanWood1 committed:
bazel lint

IanWood1 committed:
add corresponding torch test

IanWood1 committed:
file rename

IanWood1 committed:
bazel lint
